futures/src/lib.rs

```diff
@@ -192,62 +56,185 @@ macro_rules! try_ready {
+#[macro_export]
+macro_rules! task_local {
+    (static $NAME:ident: $t:ty = $e:expr) => (
+        static $NAME: $crate::task::LocalKey<$t> = {
+            fn __init() -> $t { $e }
+            fn __key() -> ::std::any::TypeId {
+                struct __A;
+                ::std::any::TypeId::of::<__A>()
+            }
+            $crate::task::LocalKey {
+                __init: __init,
+                __key: __key,
+            }
+        };
+    )
+}
 
 pub use futures_core::{Async, Poll};
 
 #[cfg(feature = "std")]
 pub mod channel {
-    //! Channels
+    //! Cross-task communciation.
+    //!
+    //! Like threads, concurrent tasks sometimes need to communicate with each
+    //! other. This module contains two basic abstractions for doing so:
     //!
-    //! This module contains channels for asynchronous
-    //! communication.
-    pub use futures_channel::*;
+    //! - [oneshot](::channel::oneshot), a way of sending a single value from
+    //! one task to another.
+    //!
+    //! - [mpsc](::channel::mpsc), a multi-producer, single-consumer channel for
+    //! sending values between tasks, analogous to the similarly-named structure
+    //! in the standard library.
+
+    pub use futures_channel::{oneshot, mpsc};
 }
 
+#[cfg(feature = "std")]
 pub mod executor {
-    //! Executors
+    //! Task execution.
+    //!
+    //! All asynchronous computation occurs within an *executor*, which is
+    //! capable of "spawning" futures as *tasks*. This module provides several
+    //! built-in executors, as well as tools for building your own.
+    //!
+    //! # Using a thread pool (M:N task scheduling)
+    //!
+    //! Most of the time tasks should be executed on a [thread
+    //! pool](::executor::ThreadPool). A small set of worker threads can handle
+    //! a very large set of spawned tasks (which are much lighter weight than
+    //! threads).
+    //!
+    //! The simplest way to use a thread pool is to
+    //! [`run`](::executor::ThreadPool::run) an initial task on it, which can
+    //! then spawn further tasks back onto the pool to complete its work:
+    //!
+    //! ```
+    //! use futures::executor::ThreadPool;
+    //! # use futures::future::{Future, lazy};
+    //! # let my_app: Box<Future<Item = (), Error = ()>> = Box::new(lazy(|| Ok(())));
     //!
-    //! This module contains the `Executor` trait, which allows futures to be
-    //! spawned and executed asynchronously.
+    //! // assumping `my_app: Future`
+    //! ThreadPool::new().run(my_app);
+    //! ```
+    //!
+    //! The call to [`run`](::executor::ThreadPool::run) will block the current
+    //! thread until the future defined by `my_app` completes, and will return
+    //! the result of that future.
+    //!
+    //! # Spawning additional tasks
+    //!
+    //! There are two ways to spawn a task:
+    //!
+    //! - Spawn onto a "default" execuctor by calling the top-level
+    //! [`spawn`](::executor::spawn) function.
```

Comment:
Possibly worth mentioning `cx.spawn(fut)` can be used when implementing futures manually. (maybe this belongs in the docs of `spawn`, though, rather than these docs?)